### PR TITLE
[Intel MKL] Adding support for convolution fusions with native format

### PR DESCRIPTION
**Pull request by mahmoud-abuzaina**

### CODE REVIEW

File: tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc [hunk 1]
```diff
@@ -715,35 +757,56 @@ class FusedPadConvOpTest : public OpsTestBase {
...
-                     .Finalize(node_def()));
-    TF_EXPECT_OK(InitOp());
+    if (!NativeFormatEnabled()) {
+      TF_EXPECT_OK(NodeDefBuilder("fused_pad_conv_op", "_MklPadWithConv2D")
+                       .Input(FakeInput(dtype))     // Input
+                       .Input(FakeInput(dtype))     // Filter
+                       .Input(FakeInput(DT_INT32))  // Padding
+                       .Input(FakeInput(DT_UINT8))  // MKl second tensor
+                       .Input(FakeInput(DT_UINT8))  // MKl second tensor
+                       .Input(FakeInput(DT_UINT8))  // MKl second tensor
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc:767]: Nit: 
```suggestion
                       .Input(FakeInput(DT_UINT8))  // MKL second tensor
                       .Input(FakeInput(DT_UINT8))  // MKL second tensor
                       .Input(FakeInput(DT_UINT8))  // MKL second tensor
```

File: tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc [hunk 2]
```diff
@@ -715,35 +757,56 @@ class FusedPadConvOpTest : public OpsTestBase {
...
+              .Input(FakeInput(dtype))     // Input
+              .Input(FakeInput(dtype))     // Filter
+              .Input(FakeInput(DT_INT32))  // Padding
+              .Attr("padding", "VALID")
+              .Attr("data_format", data_format)
+              .Attr("T", dtype)
+              .Attr("strides", {1, stride, stride, 1})
+              .Attr("_kernel", "MklNameChangeOp")
+              .Finalize(node_def()));
+      TF_EXPECT_OK(InitOp());
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc:787]: Nit: Please refactor this line out of `if/else`.

File: tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc [hunk 3]
```diff
@@ -770,33 +833,52 @@ class FilterCacheTest : public OpsTestBase {
...
-                     .Attr("T", dtype)
-                     .Attr("strides", {1, stride, stride, 1})
-                     .Attr("_kernel", "MklLayoutDependentOp")
-                     .Finalize(node_def()));
+    if (!NativeFormatEnabled()) {
+      TF_EXPECT_OK(NodeDefBuilder("conv2d_filter_cache", "_MklConv2D")
+                       .Input(FakeInput(dtype))     // Input
+                       .Input(FakeInput(dtype))     // Filter
+                       .Input(FakeInput(DT_UINT8))  // MKl second tensor
+                       .Input(FakeInput(DT_UINT8))  // MKl second tensor
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc:841]: Nit:
```suggestion
                       .Input(FakeInput(DT_UINT8))  // MKL second tensor
                       .Input(FakeInput(DT_UINT8))  // MKL second tensor
```

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 4]
```diff
@@ -61,6 +61,52 @@ REGISTER_OP("_MklFusedConv2D")
...
+    .Attr(GetConvnetDataFormatAttrString())
+    .Attr(GetExplicitPaddingsAttrString())
+    .Attr("dilations: list(int) = [1, 1, 1, 1]")
+    .Attr("fused_ops: list(string) = []")
+    // Attributes for the FusedBatchNorm ------------------------------------ //
+    .Attr("epsilon: float = 0.0001")
+    // Attributes for the LeakyRelu ----------------------------------------- //
+    .Attr("leakyrelu_alpha: float = 0.2")
+    // ---------------------------------------------------------------------- //
+    .SetShapeFn(shape_inference::Conv2DShape)
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_nn_ops.cc:83]: Maybe use `Conv2DShapeWithExplicitPadding` to match `_FusedConv2D`?
```suggestion
    .SetShapeFn(shape_inference::Conv2DShapeWithExplicitPadding)
```

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 5]
```diff
@@ -61,6 +61,52 @@ REGISTER_OP("_MklFusedConv2D")
...
+    .Doc(R"doc(
+*NOTE*: Do not invoke this operator directly in Python. MKL DNN graph transformer
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_nn_ops.cc:85]: Nit: Maybe change to oneDNN?
```suggestion
*NOTE*: Do not invoke this operator directly in Python. oneDNN graph transformer
```

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 6]
```diff
@@ -61,6 +61,52 @@ REGISTER_OP("_MklFusedConv2D")
...
+    .Attr("T: {bfloat16, float}")
+    .Attr("strides: list(int)")
+    .Attr("use_cudnn_on_gpu: bool = true")
+    .Attr("is_filter_const: bool = false")
+    .Attr(GetPaddingAttrString())
+    .Attr(GetConvnetDataFormatAttrString())
+    .Attr("dilations: list(int) = [1, 1, 1, 1]")
+    .SetShapeFn(shape_inference::Conv2DShape)
+    .Doc(R"doc(
+MKL version of Conv2D and BiasAdd operator. Uses DNNL APIs to perform
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_nn_ops.cc:103]: Nit:
```suggestion
MKL version of Conv2D and BiasAdd operator. Uses oneDNN APIs to perform
```

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 7]
```diff
@@ -169,6 +235,55 @@ REGISTER_OP("_MklPadWithFusedConv2D")
...
+    .Attr("fused_ops: list(string) = []")
+    .Attr("Tpaddings: {int32, int64} = DT_INT32")
+    // Attributes for the FusedBatchNorm ------------------------------------ //
+    .Attr("epsilon: float = 0.0001")
+    // Attributes for the LeakyRelu ----------------------------------------- //
+    .Attr("leakyrelu_alpha: float = 0.2")
+    // ---------------------------------------------------------------------- //
+    .SetShapeFn(shape_inference::Conv2DShape)
+    .Doc(R"doc(
+*NOTE*: Do not invoke this operator directly in Python. MKL DNN graph transformer
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_nn_ops.cc:260]: Nit:
```suggestion
*NOTE*: Do not invoke this operator directly in Python. oneDNN graph transformer
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and sorry for my delay!

File: tensorflow/core/kernels/mkl/mkl_conv_ops.cc [hunk 8]
```diff
@@ -937,8 +937,20 @@ class MklConvOp : public OpKernel {
...
+      if (native_format) {
+        // Forward the summand tensor to the output only if it has no other
+        // references, otherwise make a copy of it.
+        if (!context->forward_input_to_output_with_shape(
+                kInputIndex_Add, kOutputIndex_Dst, output_tf_shape,
+                output_tensor)) {
+          AllocateOutputSetMklShape(context, kOutputIndex_Dst, output_tensor,
+                                    output_tf_shape, *output_mkl_shape,
+                                    native_format);
+          CHECK((*output_tensor)->CopyFrom(add_tensor, add_tensor.shape()));
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_conv_ops.cc]: Fixed.

File: tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc (hunk 1 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc]: Done.

File: tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc (hunk 2 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc]: Done.

File: tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc (hunk 3 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_fused_ops_test.cc]: Done.

File: tensorflow/core/ops/mkl_nn_ops.cc (hunk 4 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Done.

File: tensorflow/core/ops/mkl_nn_ops.cc (hunk 5 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Done.

File: tensorflow/core/ops/mkl_nn_ops.cc (hunk 6 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Done.

File: tensorflow/core/ops/mkl_nn_ops.cc (hunk 7 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Done.

**Comment by mahmoud-abuzaina** [comment]
Thank you for reviewing the PR. I have addressed your comments.

**Review by penpornk — APPROVED** [review]
Thank you again for the PR!

**Pull request merged**